syne_tune/optimizer/schedulers/botorch/botorch_searcher.py

```diff
@@ -0,0 +1,235 @@
+                not_enough_suggestion = len(self.objectives()) < self.num_minimum_observations
+                if not_enough_suggestion:
+                    config_suggested = self._sample_random()
+                else:
+                    config_suggested = self._sample_next_candidate()
+
+        self.pending_trials.add(trial_id)
+        self.trial_configs[trial_id] = config_suggested
+        self.config_seen.add(tuple(config_suggested.values()))
+
+        return config_suggested
+
+    def _sample_next_candidate(self) -> Dict:
+        """
+        :return: a next candidate to evaluate, if possible it is obtained by fitting a GP on past data and maximizing EI
+        if this fails because of numerical difficulties with non PSD matrices, then the candidate is sampled at random.
+        """
+        try:
+            X = np.array(self._config_to_feature_matrix(self._configs_with_results()))
+            y = self.objectives()
+
+            if self.max_num_observations is not None and len(X) >= self.max_num_observations:
+                perm = self.random_state.permutation(len(X))
+                X = X[perm]
+                y = y[perm]
+                subsample = True
+            else:
+                subsample = False
+
+            X_tensor = torch.Tensor(X)
+            Y_tensor = standardize(torch.Tensor(y).reshape(-1, 1))
+            gp = self._make_gp(X_tensor=X_tensor, Y_tensor=Y_tensor)
+            mll = ExactMarginalLogLikelihood(gp.likelihood, gp)
+            fit_gpytorch_model(mll, max_retries=0)
+
+            if self.pending_trials and self.fantasising and not subsample:
+                X_pending = self._config_to_feature_matrix(self._configs_pending())
+            else:
+                X_pending = None
+
+            acq = qExpectedImprovement(
+                model=gp,
+                best_f=Y_tensor.min().item() if self.mode == 'min' else Y_tensor.max().item(),
+                maximize=self.mode == 'max',
+                X_pending=X_pending,
+            )
+
+            candidate, acq_value = optimize_acqf(
+                acq,
+                bounds=torch.Tensor(self.hp_ranges.get_ndarray_bounds()).T,
+                q=1,
+                num_restarts=3,
+                raw_samples=100
+            )
+
+            candidate = candidate.detach().numpy()[0]
+            config = self.hp_ranges.from_ndarray(candidate)
+            if not self._is_config_already_seen(config):
+                return config
+            else:
+                return self._sample_and_pick_acq_best(acq)
+        except NotPSDError as _:
+            logging.info("Chlolesky inversion failed, sampling randomly.")
+            return self._sample_random()
+
+    def _make_gp(self, X_tensor: torch.Tensor, Y_tensor: torch.Tensor) -> SingleTaskGP:
+        double_precision = False
+        if double_precision:
+            X_tensor = X_tensor.double()
+            Y_tensor = Y_tensor.double()
+
+        noise_std = 1e-3
```

Comment:
Why do we no learn this hyperparameter together with the other hyperparameter of the GP? Could we make it at least an input argument?

Reply:
It somehow can give numerical instability, I set it to a constant as done in other GP searchers.